### PR TITLE
WIP Add Managed Zones

### PR DESCRIPTION
**Pull request by mikenairn**
ManagedZones replaces the hard coded controller DNS Provider configuration and DNSZone with a new ManagedZone CRD. ManagedZone has a `rootDomain` and holds the DNSProvider configuration to create DNS records in the remote provider using that domain (currently only supports Route53). A ManagedZone references a secret that contains the AWS credentials for the provider and both a ManagedZone and it's corresponding credentials must be created in order for mctc to successfully reconcile DNSRecords into route53.

DNSRecords are assigned a single ManagedZone on initial creation. Currently mctc will look for a ManagedZone in the DNSRecords namespace, but if none are found it will look in an mctc designated namespace for default ManagedZone to use, default ns is "default", but can be overriden with `--managed-zones-namespace mctc`. This is very basic logic at the moment but should be expanded in future iterations.

```
make local-setup
make build install && bin/manager --managed-zones-namespace mctc --zap-log-level=3
```

```
#!/bin/bash

kubectl delete dnsrecords --all -A
kubectl delete managedzones --all -A

kubectl delete ns test-dev-hcapps-net --wait=true
kubectl delete ns test-mn-hcapps-net --wait=true
kubectl delete ns mctc --wait=true

# Namespace with a managed zone for (dev.hcapps.net)
kubectl create ns mctc
kubectl apply -f kuadrant.io_v1_managedzone_dev.hcapps.net.yaml -n mctc
kubectl create secret generic aws-credentials --from-env-file=aws-credentials-dev.hcapps.net.env -n mctc

# Namespace with no managed zone, and a dns record for (dev.hcapps.net)
kubectl create ns test-dev-hcapps-net
kubectl apply -f kuadrant.io_v1_dnsrecord_dev.hcapps.net.yaml -n test-dev-hcapps-net

# Namespace with a managed zone for (mn.hcapps.net), and a dns record for (mn.hcapps.net)
kubectl create ns test-mn-hcapps-net
kubectl apply -f kuadrant.io_v1_managedzone_mn.hcapps.net.yaml -n test-mn-hcapps-net
kubectl create secret generic aws-credentials --from-env-file=aws-credentials-mn.hcapps.net.env -n test-mn-hcapps-net
kubectl apply -f kuadrant.io_v1_dnsrecord_mn.hcapps.net.yaml -n test-mn-hcapps-net
```
Note: Above requires appropriate managedzone, aws credentials and dnsrecord resources.

Examples:

kuadrant.io_v1_dnsrecord_dev.hcapps.net.yaml
```
apiVersion: kuadrant.io/v1
kind: ManagedZone
metadata:
  name: dev.hcapps.net
spec:
  rootDomain: "dev.hcapps.net"
  route53:
    accessKeyID: "My AWS ACCESS KEY"
    hostedZoneID: "MY HOSTED ZONE ID"
    region: "us-east-1"
    secretAccessKeySecretRef:
      name: "aws-credentials"
      key: "AWS_SECRET_ACCESS_KEY"
```

aws-credentials-dev.hcapps.net.env
```
AWS_SECRET_ACCESS_KEY=MY_AWS_SECRET_ACCESS_KEY
```

kuadrant.io_v1_dnsrecord_dev.hcapps.net.yaml
```
apiVersion: kuadrant.io/v1
kind: DNSRecord
metadata:
  name: dnsrecord-sample
spec:
  endpoints:
    - dnsName: dnsrecord-simple.dev.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:
        - 52.215.108.61
        - 52.30.101.221
```
 

```
$ kubectl get managedzones -A
NAMESPACE            NAME             DOMAIN
mctc                 dev.hcapps.net   dev.hcapps.net
test-mn-hcapps-net   mn.hcapps.net    mn.hcapps.net
$ kubectl get dnsrecords -A
NAMESPACE             NAME               MZ NAME          MZ NAMESPACE         READY
test-dev-hcapps-net   dnsrecord-sample   dev.hcapps.net   mctc                 True
test-mn-hcapps-net    dnsrecord-sample   mn.hcapps.net    test-mn-hcapps-net   True
```

### CODE REVIEW

File: pkg/apis/v1/dnsrecord_types.go [hunk 1]
```diff
@@ -123,10 +130,45 @@ type DNSRecordStatus struct {
...
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// endpoints are the last endpoints that were successfully published by the provider
+	//
+	// Provides a simple mechanism to store the current provider records in order to
+	// delete any that are no longer present in DNSRecordSpec.Endpoints
+	//
+	// Note: This will not be required if/when we switch to using external-dns since when
+	// running with a "sync" policy it will clean up unused records automatically.
+	Endpoints []*Endpoint `json:"endpoints,omitempty"`
```
Comment (maleck13) [inline pkg/apis/v1/dnsrecord_types.go:141]: did we want to also store the nameservers for this zone in the status? So if a user wants to delegate a zone to us we can provide the needed nameservers for them to update their own DNS zone with NS records

Comment (maleck13) [reply pkg/apis/v1/dnsrecord_types.go:141]: oh nvm this is the DNSRecord

File: pkg/apis/v1/managedzone_types.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	SecretAccessKey cmmeta.SecretKeySelector `json:"secretAccessKeySecretRef"`
+
+	HostedZoneID string `json:"hostedZoneID"`
+
+	// Always set the region when using AccessKeyID and SecretAccessKey
+	Region string `json:"region"`
+}
+
+// ManagedZoneStatus defines the observed state of ManagedZone
+type ManagedZoneStatus struct {
```
Comment (maleck13) [inline pkg/apis/v1/managedzone_types.go:52]: ok so in this resource is where we want to return the assigned nameservers

**Comment by maleck13** [comment]
Ok  so I kinda expected to find logic in here to create the new zone.  So I guess currently we expect the zone to exist and creating the zone resource just gives the controller perms to use that zone?

seems like we have taken the first step which is to allow the controller to use more than one zone, so I guess the natural follow on here is to add logic to allow the controller to setup the zone and (possibly) a new set of credentials for accessing just that zone?

**Comment by mikenairn** [comment]
Closing for the moment

**Pull request closed without merging**